packages/core/src/internal/clients/leaderboard/AbstractLeaderboardClient.ts

```diff
@@ -0,0 +1,199 @@
+  LeaderboardFetch,
+  LeaderboardGetRank,
+  LeaderboardLength,
+  LeaderboardRemoveElements,
+  LeaderboardUpsert,
+  SortedSetOrder,
+} from '../../../index';
+import {
+  ILeaderboardClient,
+  LeaderboardFetchByRankOptions,
+  LeaderboardFetchByScoreOptions,
+} from '../../../clients/ILeaderboardClient';
+import {InternalLeaderboardClient} from './InternalLeaderboardClient';
+
+export interface BaseLeaderboardClientProps {
+  createLeaderboardClient: () => InternalLeaderboardClient;
+}
+export abstract class AbstractLeaderboardClient implements ILeaderboardClient {
+  // making these protected until we fully abstract away the nodejs client
+  protected readonly leaderboardClient: InternalLeaderboardClient;
+
+  protected constructor(props: BaseLeaderboardClientProps) {
+    this.leaderboardClient = props.createLeaderboardClient();
+  }
+
+  /**
+   * Updates elements in a leaderboard or inserts elements if they do not already exist.
+   * The leaderboard is also created if it does not already exist.
+   *
+   * @param {string} cacheName - The cache containing the leaderboard.
+   * @param {string} leaderboardName - The leaderboard to upsert to.
+   * @param {Map<bigint, number>} elements - The ID->score pairs to add to the leaderboard.
+   * @returns {Promise<LeaderboardUpsert.Response>} -
+   * {@link LeaderboardUpsert.Success} on success.
+   * {@link LeaderboardUpsert.Error} on failure.
+   */
+  public async leaderboardUpsert(
+    cacheName: string,
+    leaderboardName: string,
+    elements: Map<bigint, number>
+  ): Promise<LeaderboardUpsert.Response> {
+    return await this.leaderboardClient.leaderboardUpsert(
+      cacheName,
+      leaderboardName,
+      elements
+    );
+  }
+
+  /**
+   * Fetch the elements in the given leaderboard by score.
+   *
+   * @param {string} cacheName - The cache containing the leaderboard.
+   * @param {string} leaderboardName - The leaderboard to fetch from.
+   * @param {LeaderboardFetchByScoreOptions} options
+   * @param {number} [options.minScore] - The minimum score (inclusive) of the
+   * elements to fetch. Defaults to negative infinity.
+   * @param {number} [options.maxScore] - The maximum score (inclusive) of the
+   * elements to fetch. Defaults to positive infinity.
+   * @param {SortedSetOrder} [options.order] - The order to fetch the elements in.
+   * Defaults to ascending.
+   * @param {number} [options.offset] - The number of elements to skip before
+   * returning the first element. Defaults to 0. Note: this is not the score of
+   * the first element to return, but the number of elements of the result set
+   * to skip before returning the first element.
+   * @param {number} [options.count] - The maximum number of elements to return.
+   * Defaults to undefined, which returns all elements.
+   * @returns {Promise<LeaderboardFetch.Response>} -
+   * {@link LeaderboardFetch.Success} containing the requested elements.
+   * {@link LeaderboardFetch.Error} on failure.
+   */
```

Comment:
similar comment here - this function has the potential to either blow out the memory on the client, or get throttled by the server, for large leaderboards. We will need to be able to advise customers on how to use it. If backend team doesn't have concrete answers for now then we need a ticket to circle back to this and add link to docs/examples.